client/factory.go

```diff
@@ -0,0 +1,188 @@
+package client
+
+import (
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/k8sclient"
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/azure-operator/v4/pkg/credential"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+type FactoryConfig struct {
+	K8sClient  k8sclient.Interface
+	GSTenantID string
+}
+
+// Factory is creating Azure clients for specified AzureConfig CRs, so basically for specified
+// tenant clusters. All created clients are cached.
+type Factory struct {
+	mutex      sync.Mutex
+	k8sClient  k8sclient.Interface
+	gsTenantID string
+
+	clients map[string]*AzureClientSet
+}
+
+type clientCreatorFunc func(autorest.Authorizer, string, string) (interface{}, error)
+
+// NewFactory returns a new Azure client factory that is used throughout entire azure-operator
+// lifetime.
+func NewFactory(config FactoryConfig) (*Factory, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if len(config.GSTenantID) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GSTenantID must not be empty", config)
+	}
+
+	factory := &Factory{
+		k8sClient:  config.K8sClient,
+		gsTenantID: config.GSTenantID,
+		clients:    make(map[string]*AzureClientSet),
+	}
+
+	return factory, nil
+}
+
+// GetDeploymentsClient returns DeploymentsClient that is used for management of deployments and
+// ARM templates. The client (for specified cluster) is cached after creation, so the same client
+// is returned every time.
+func (f *Factory) GetDeploymentsClient(cr v1alpha1.AzureConfig) (*resources.DeploymentsClient, error) {
+	clientSetKey := key.CredentialName(cr)
+
+	f.mutex.Lock()
+	if _, ok := f.clients[clientSetKey]; !ok {
+		f.clients[clientSetKey] = &AzureClientSet{}
+	}
+	if f.clients[clientSetKey].DeploymentsClient == nil {
+		deploymentClient, err := f.createClient(cr, func(authorizer autorest.Authorizer, subscriptionID string, partnerID string) (interface{}, error) {
+			return newDeploymentsClient(authorizer, subscriptionID, partnerID)
+		})
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		f.clients[clientSetKey].DeploymentsClient = deploymentClient.(*resources.DeploymentsClient)
```

Comment:
Creating clients in a generic manner by returning `interface{}` from `createClient(...)` and casting back to actual client type, so there is no too much code repetition.

Performance overhead should not be a concern, as this is done only once for every client, over the lifetime of the `azure-operator`.

Reply:
P.S. I miss generics :)